### PR TITLE
DRV-59: Added ability to use custom HttpClient - fixes issue #73 

### PR DESCRIPTION
**Pull request by kcrandall**
See issue #73 for link as to why this is important. Simple fix to allow library user to supply their own HttpClient

### CODE REVIEW

File: FaunaDB.Client/Client/DefaultClientIO.cs [hunk 1]
```diff
@@ -29,9 +32,12 @@ internal DefaultClientIO(HttpClient client, AuthenticationHeaderValue authHeader
             this.lastSeen = lastSeen;
         }
 
-        public DefaultClientIO(string secret, Uri endpoint, TimeSpan timeout)
-            : this(CreateClient(endpoint, timeout), AuthHeader(secret), new LastSeen())
-        { }
+        public DefaultClientIO(string secret, Uri endpoint, TimeSpan timeout, HttpClient httpClient = null)
+            : this(CreateClient(endpoint, timeout, httpClient), AuthHeader(secret), new LastSeen())
```
Comment (marrony) [inline FaunaDB.Client/Client/DefaultClientIO.cs]: no need to pass `httpClient` to the function `CreateClient`, if you already have it, just ignore the function call.
```csharp
httpClient ?? CreateClient(endpoint, timeout)
```

File: FaunaDB.Client/Client/DefaultClientIO.cs [hunk 2]
```diff
@@ -117,17 +128,16 @@ static string QueryString(IReadOnlyDictionary<string, string> query)
             return string.Join("&", keyValues);
         }
 
-        static HttpClient CreateClient(Uri endpoint, TimeSpan timeout)
+        static HttpClient CreateClient(Uri endpoint, TimeSpan timeout, HttpClient httpClient)
```
Comment (marrony) [inline FaunaDB.Client/Client/DefaultClientIO.cs]: remove `HttpClient httpClient`

File: FaunaDB.Client/Client/DefaultClientIO.cs [hunk 3]
```diff
@@ -48,9 +54,14 @@ async Task<RequestResult> DoRequestAsync(HttpMethodKind method, string path, str
 
             var startTime = DateTime.UtcNow;
 
-            var message = new HttpRequestMessage(new HttpMethod(method.Name()), path);
+            var message = new HttpRequestMessage(new HttpMethod(method.Name()), $"{endpoint}{path}");
```
Comment (marrony) [inline FaunaDB.Client/Client/DefaultClientIO.cs]: why did you add `endpoint` here? the client already has the endpoint as the `BaseAddress`

File: FaunaDB.Client/Client/DefaultClientIO.cs [hunk 4]
```diff
@@ -48,9 +54,14 @@ async Task<RequestResult> DoRequestAsync(HttpMethodKind method, string path, str
...
             message.Content = dataString;
             message.Headers.Authorization = authHeader;
+            message.Headers.AcceptEncoding.Add(new StringWithQualityHeaderValue("gzip"));
+            message.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            message.Headers.Add("X-FaunaDB-API-Version", "2.7");
+            message.Headers.Add("X-Fauna-Driver", "csharp");
+            message.SetTimeout(timeout);
```
Comment (marrony) [inline FaunaDB.Client/Client/DefaultClientIO.cs]: why are you setting timeout here? we are already setting this on `HttpClient` creation

File: FaunaDB.Client/Client/DefaultClientIO.cs (hunk 3 above)
Comment (kcrandall) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: If the user is passing their own then they might not have a base address set and you don't want to change theirs or set one

File: FaunaDB.Client/Client/DefaultClientIO.cs (hunk 4 above)
Comment (kcrandall) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: I removed that, because the user may pass their own. So you need to set it on the HttpRequestMessage for each call. The user may have their own timeout or the default. Don't want to change it and mess up how their other requests deal with them

File: FaunaDB.Client/Client/DefaultClientIO.cs (hunk 1 above)
Comment (kcrandall) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: I need to pass the httpClient if the user passes their own in FaunaClient(httpClient: userDefinedClient)

File: FaunaDB.Client/Client/DefaultClientIO.cs (hunk 2 above)
Comment (kcrandall) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: have to pass it from FaunaClient() constructor

File: FaunaDB.Client/Client/DefaultClientIO.cs (hunk 1 above)
Comment (kcrandall) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: wait I see what you are saying on this one

Comment (kcrandall) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: Ya we can ignore the call too I guess

File: FaunaDB.Client/Client/DefaultClientIO.cs (hunk 2 above)
Comment (kcrandall) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: and this can change as a result of the last one

File: FaunaDB.Client/Client/DefaultClientIO.cs (hunk 3 above)
Comment (marrony) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: can you add a test where you create your own `HttpClient` not specifying `BaseAddress` on it, just to make sure we are handling the `endpoint` + `path` correctly

Comment (kcrandall) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: sure what test file do you want it in?

Comment (kcrandall) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: ClientTest.cs?

Comment (kcrandall) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: How do I run unit tests that require networking? they say connection refused do I need to supply a secret key somewhere?

Comment (kcrandall) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: The client should act the same though with user supplied or without because even if the client creates its own its not setting a base. So even without user supplied httpclient if it doesn't work all the current tests will fail

Comment (marrony) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: pretty much all tests extends this https://github.com/fauna/faunadb-csharp/blob/master/FaunaDB.Client.Test/TestCase.cs#L42 so you can set an env variable with your key/secret and domain/port.

Anyway, without a test I cannot ensure that a custom httpclient set with a BaseAddress will not interfere with endpoint+path

File: FaunaDB.Client/Client/DefaultClientIO.cs [hunk 5]
```diff
@@ -120,12 +131,8 @@ static string QueryString(IReadOnlyDictionary<string, string> query)
         static HttpClient CreateClient(Uri endpoint, TimeSpan timeout)
         {
             var client = new HttpClient();
-            client.BaseAddress = endpoint;
-            client.Timeout = timeout;
-            client.DefaultRequestHeaders.AcceptEncoding.Add(new StringWithQualityHeaderValue("gzip"));
-            client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
-            client.DefaultRequestHeaders.Add("X-FaunaDB-API-Version", "2.7");
-            client.DefaultRequestHeaders.Add("X-Fauna-Driver", "csharp");
+            //client.BaseAddress = endpoint;
```
Comment (marrony) [inline FaunaDB.Client/Client/DefaultClientIO.cs]: remove commented code

File: FaunaDB.Client/Client/DefaultClientIO.cs [hunk 6]
```diff
@@ -47,10 +53,15 @@ async Task<RequestResult> DoRequestAsync(HttpMethodKind method, string path, str
                 path = $"{path}?{queryString}";
 
             var startTime = DateTime.UtcNow;
-
-            var message = new HttpRequestMessage(new HttpMethod(method.Name()), path);
+            System.Diagnostics.Debug.WriteLine($"Endpoint: {endpoint}{path}");
```
Comment (marrony) [inline FaunaDB.Client/Client/DefaultClientIO.cs]: don't forget to remove this

File: FaunaDB.Client/Client/DefaultClientIO.cs (hunk 3 above)
Comment (marrony) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: Or you can set the base address of the custom httpclient to a non related address and ensure we are doing requests to fauna address specified on endpoint

Comment (kcrandall) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: Ok i passed an httpclient in TestCase with a random baseaddress. it worked as expected and the tests passed. HttpRequestMessage will always override the base address if a domain is passed to it

**Comment by evbruno** [comment]
followup ticket: DRV-99

File: FaunaDB.Client/Client/FaunaClient.cs [hunk 7]
```diff
@@ -33,8 +33,9 @@ public class FaunaClient
         public FaunaClient(
             string secret,
             string endpoint = "https://db.fauna.com:443",
-            TimeSpan? timeout = null)
-            : this(CreateClient(secret, endpoint, timeout))
+            TimeSpan? timeout = null,
+            System.Net.Http.HttpClient httpClient = null)
```
Comment (marrony) [inline FaunaDB.Client/Client/FaunaClient.cs:37]: could you add the import for this class in the top of the file and remove the full qualified name here?

File: FaunaDB.Client/Client/HttpRequestExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
+﻿using System;
+using System.Net.Http;
+namespace FaunaDB.Client
+{
+    public static class HttpRequestExtensions
+    {
+        private static string TimeoutPropertyKey = "RequestTimeout";
```
Comment (marrony) [inline FaunaDB.Client/Client/HttpRequestExtensions.cs:7]: Where did you find oficial documentation about this property? Microsoft documentation don't mention anything about it https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httprequestmessage.properties?view=netframework-4.8

Comment (kcrandall) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs:7]: https://thomaslevesque.com/2018/02/25/better-timeout-handling-with-httpclient/

Comment (kcrandall) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs:7]: this is another alternative to set cancellation token https://stackoverflow.com/questions/28060575/how-to-set-time-out-for-http-client-request-operation-in-windows-phone-8-1-windo

Comment (kcrandall) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs:7]: I can add the cancellation token too if that's better

Comment (marrony) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs:7]: looks like `request.Properties` is not used in any place on the source code https://github.com/microsoft/referencesource/tree/master/System/net/System/Net/Http

Comment (kcrandall) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs:7]: looks like the best thing to do is this. It will cancel the request when the token expires causing it timeout 
```
CancellationTokenSource cts = new CancellationTokenSource(2000); // 2 seconds
    HttpClient client = new HttpClient();
    HttpResponseMessage response = await
        client.SendRequestAsync(request).AsTask(cts.Token);
```

Comment (evbruno) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs:7]: I think we could remove the extension, and add the timeout in the client creation... I can confirm that this patch works:
```
diff --git a/FaunaDB.Client/Client/DefaultClientIO.cs b/FaunaDB.Client/Client/DefaultClientIO.cs
index 7e1f3b7..0e5fd0c 100644
--- a/FaunaDB.Client/Client/DefaultClientIO.cs
+++ b/FaunaDB.Client/Client/DefaultClientIO.cs
@@ -35,7 +35,7 @@ namespace FaunaDB.Client
         }

         public DefaultClientIO(string secret, Uri endpoint, TimeSpan timeout, HttpClient httpClient = null)
-            : this(httpClient ?? CreateClient(), AuthHeader(secret), new LastSeen(), endpoint, timeout)
+            : this(httpClient ?? CreateClient(timeout), AuthHeader(secret), new LastSeen(), endpoint, timeout)
         { }

         public IClientIO NewSessionClient(string secret) =>
@@ -127,9 +127,14 @@ namespace FaunaDB.Client
             return string.Join("&", keyValues);
         }

-        static HttpClient CreateClient()
+        static HttpClient CreateClient(TimeSpan? timeout)
         {
-            return new HttpClient();
+            var c = new HttpClient();
+
+            if (timeout.HasValue)
+                c.Timeout = timeout.Value;
+
+            return c;
         }
     }
 }```

Comment (evbruno) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs:7]: Ok... you can remove every timeout related code, we are going to manage this in the server side.

Comment (kcrandall) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs:7]: I assume you'll need to send timeout in some header or something so ill just remove the extensions and SetTimeout() call in DefaultClientIO and leave timeout in the DefaultClientIO constructor?

Comment (marrony) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs]: yeah, that's the idea, the driver will send a header with the desired timeout per query and it will control the entire pipeline on the server side

Comment (kcrandall) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs]: ok so the new commit should fix that you guys just have to add the header in a separate commit

Comment (kcrandall) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs]: if you want to give me the header key I can set it to the timeout variable

Comment (evbruno) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs]: Before we move to this new header approach, I would like to finish and merge this PR. Could  you do it ? We just need to remove the current timeout related stuff and this PR is good to go!

Comment (evbruno) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs]: In the meantime I'm looking for the header specs (such as, is this milliseconds or seconds?)

Comment (kcrandall) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs]: Wouldn't the best way to do it server side is have two headers one is an int. The other is an enum with values mil, sec, tick etc. Then it doesn't matter what the clients send users can send whatever they want and server side figures out the conversions

Comment (kcrandall) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs]: I don't off the top of my head know what C# timeout is bc I always leave it default

Comment (kcrandall) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs]: I removed the timeout stuff too in commit f0a1c92. It still exists in the class ready to be put in a header

Comment (kcrandall) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs]: the client uses TimeSpan so users can set it to anything TimeSpan.FromSeconds()/Hours() etc

Comment (kcrandall) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs]: you can do this on the client timeout.TotalMilliseconds to then pass it to the header

Comment (evbruno) [reply FaunaDB.Client/Client/HttpRequestExtensions.cs]: I'm not sure YET how the header is going to behave but using TimeSpan to convert to milliseconds or seconds sounds really nice (user friendly)
